File: backend/app/routers/recycle.py
# -*- coding: utf-8 -*-
"""系统回收站与业务审计接口。"""
from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel

from ..services import audit, recycle

router = APIRouter(prefix='/api')


class PurgeBody(BaseModel):
    confirmation: str = ''


def _error(exc: recycle.RecycleError):
    raise HTTPException(400, str(exc)) from exc


def _is_local_host(host: str) -> bool:
    return host in {'127.0.0.1', '::1', 'localhost', 'testclient'}


@router.get('/recycle-bin')
def list_recycle_bin(object_type: str = '', status: str = '已删除',
                     limit: int = Query(300, ge=1, le=500)):
    return {'items': recycle.list_entries(
        object_type=object_type, status=status, limit=limit)}


@router.delete('/records/{object_type}/{object_id}')
def delete_record(object_type: str, object_id: int):
    try:
        return recycle.soft_delete(object_type, object_id)
    except recycle.RecycleError as exc:
        _error(exc)


@router.post('/recycle-bin/{entry_id}/restore')
def restore_record(entry_id: int):
    try:
        return recycle.restore(entry_id)
    except recycle.RecycleError as exc:
        _error(exc)


@router.delete('/recycle-bin/{entry_id}/purge')
def purge_record(entry_id: int, body: PurgeBody, request: Request):
    host = request.client.host if request.client else ''
    if not _is_local_host(host):
        raise HTTPException(403, '永久删除只能在工作台本机操作')
    try:
        return recycle.purge(entry_id, body.confirmation)
    except recycle.RecycleError as exc:
        _error(exc)


@router.get('/system/audit')
def list_system_audit(limit: int = Query(200, ge=1, le=500)):
    return {'items': audit.list_audits(limit=limit)}

File: server/src/http/routes/recycle.ts
/* 回收站与审计路由（与 backend/app/routers/recycle.py 对应）。 */
import type { FastifyInstance, FastifyReply } from 'fastify';

import * as recycle from '../../services/recycle.js';

function isLocalHost(host: string): boolean {
  return ['127.0.0.1', '::1', 'localhost', 'testclient'].includes(host);
}

function mapError(reply: FastifyReply, error: unknown): FastifyReply | undefined {
  if (error instanceof recycle.RecycleError) {
    return reply.status(400).send({ detail: (error as Error).message });
  }
  return undefined;
}

function wrap(reply: FastifyReply, fn: () => unknown): unknown {
  try {
    return fn();
  } catch (error) {
    const mapped = mapError(reply, error);
    if (mapped) return mapped;
    throw error;
  }
}

export function registerRecycleRoutes(app: FastifyInstance): void {
  app.get('/api/recycle-bin', async (request) => {
    const { object_type = '', status = '已删除', limit = 300 } = request.query as {
      object_type?: string; status?: string; limit?: string | number;
    };
    return {
      items: recycle.listEntries({
        objectType: String(object_type),
        status: String(status),
        limit: Number(limit),
      }),
    };
  });

  app.delete('/api/records/:objectType/:objectId', async (request, reply) => {
    const { objectType, objectId } = request.params as { objectType: string; objectId: string };
    return wrap(reply, () => recycle.softDelete(objectType, Number(objectId)));
  });

  app.post('/api/recycle-bin/:entryId/restore', async (request, reply) => {
    const { entryId } = request.params as { entryId: string };
    return wrap(reply, () => recycle.restore(Number(entryId)));
  });

  app.delete('/api/recycle-bin/:entryId/purge', async (request, reply) => {
    const { entryId } = request.params as { entryId: string };
    if (!isLocalHost(request.ip)) {
      return reply.status(403).send({ detail: '永久删除只能在工作台本机操作' });
    }
    const body = (request.body ?? {}) as { confirmation?: string };
    return wrap(reply, () => recycle.purge(Number(entryId), String(body.confirmation ?? '')));
  });
}
